packages/core/core/generate-text/generate-text.ts

```diff
@@ -107,77 +108,87 @@ By default, it's set to 0, which will disable the feature.
+          toolCalls: currentToolCalls,
+          toolResults: currentToolResults,
+        });
+        responseMessages.push(...newResponseMessages);
+        promptMessages.push(
+          ...newResponseMessages.map(convertToLanguageModelMessage),
+        );
+      } while (
+        // there are tool calls:
+        currentToolCalls.length > 0 &&
+        // all current tool calls have results:
+        currentToolResults.length === currentToolCalls.length &&
+        // the number of roundtrips is less than the maximum:
+        roundtrips++ < maxToolRoundtrips
+      );
+
+      return new GenerateTextResult({
+        // Always return a string so that the caller doesn't have to check for undefined.
+        // If they need to check if the model did not return any text,
+        // they can check the length of the string:
+        text: currentModelResponse.text ?? '',
+        toolCalls: currentToolCalls,
+        toolResults: currentToolResults,
+        finishReason: currentModelResponse.finishReason,
+        usage: calculateTokenUsage(currentModelResponse.usage),
+        warnings: currentModelResponse.warnings,
+        rawResponse: currentModelResponse.rawResponse,
+        logprobs: currentModelResponse.logprobs,
+        responseMessages,
       });
-    });
-
-    // parse tool calls:
-    currentToolCalls = (currentModelResponse.toolCalls ?? []).map(
-      modelToolCall => parseToolCall({ toolCall: modelToolCall, tools }),
-    );
-
-    // execute tools:
-    currentToolResults =
-      tools == null
-        ? []
-        : await executeTools({ toolCalls: currentToolCalls, tools });
-
-    // append to messages for potential next roundtrip:
-    const newResponseMessages = toResponseMessages({
-      text: currentModelResponse.text ?? '',
-      toolCalls: currentToolCalls,
-      toolResults: currentToolResults,
-    });
-    responseMessages.push(...newResponseMessages);
-    promptMessages.push(
-      ...newResponseMessages.map(convertToLanguageModelMessage),
-    );
-  } while (
-    // there are tool calls:
-    currentToolCalls.length > 0 &&
-    // all current tool calls have results:
-    currentToolResults.length === currentToolCalls.length &&
-    // the number of roundtrips is less than the maximum:
-    roundtrips++ < maxToolRoundtrips
-  );
-
-  return new GenerateTextResult({
-    // Always return a string so that the caller doesn't have to check for undefined.
-    // If they need to check if the model did not return any text,
-    // they can check the length of the string:
-    text: currentModelResponse.text ?? '',
-    toolCalls: currentToolCalls,
-    toolResults: currentToolResults,
-    finishReason: currentModelResponse.finishReason,
-    usage: calculateTokenUsage(currentModelResponse.usage),
-    warnings: currentModelResponse.warnings,
-    rawResponse: currentModelResponse.rawResponse,
-    logprobs: currentModelResponse.logprobs,
-    responseMessages,
+    } finally {
+      span.end();
```

Comment:
It really helps to also do `catch() { span.recordException(error); span.setStatus(error)}`.